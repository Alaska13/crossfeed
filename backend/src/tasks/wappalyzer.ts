import { Handler } from 'aws-lambda';
import { connectToDatabase, Domain, Service, WebInfo } from '../models';
import { saveServicesToDb, saveWebInfoToDb } from './helpers';
import { plainToClass } from 'class-transformer';
import { Like } from 'typeorm';
import * as wappalyzer from 'simple-wappalyzer';
import axios from 'axios';
import { length } from 'class-validator';
import { catchClause } from '@babel/types';

export const handler: Handler = async (event) => {
  await connectToDatabase();

  const domains = await Domain.find({
    where: { reverseName: Like('gov.cisa%') },
    relations: ['services']
  });
  const services: Service[] = [];
  let count = 0;
  for (const domain of domains) {
    const ports = domain.services.map((s) => s.port);
    if (!ports.includes('80') && !ports.includes('443')) continue; // Change this to filter at query level
    const url = (ports.includes('443') ? 'https://' : 'http://') + domain.name;
    try {
      const { data, status, headers } = await axios.get(url);
      const result = await wappalyzer({ url, data, status, headers });
      console.log(result);
      if (result.length == 0) continue;
      const frameworkString = result.map((r) => r.name).join(', '); // Figure out how to better store this, should separate based on category
      await saveWebInfoToDb(
        plainToClass(WebInfo, {
          domain: domain,
          frameworks: frameworkString
        })
      );
      count++;
    } catch {
      continue;
    }
  }

  console.log(`Wappalyzer updated for ${count} new services`);
};
