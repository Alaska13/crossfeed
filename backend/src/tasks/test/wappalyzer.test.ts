import wappalyzer from "../wappalyzer";
jest.mock("../helpers/getLiveWebsites");
jest.mock("../helpers/saveDomainsToDb");

describe("wappalyzer", () => {
  test("basic test", async () => {
    await wappalyzer({
      organizationId: "organizationId",
      organizationName: "organizationName",
      scanId: "scanId",
      scanName: "scanName"
    })
  })
})