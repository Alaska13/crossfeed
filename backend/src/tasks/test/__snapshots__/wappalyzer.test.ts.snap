// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wappalyzer basic test: helpers.saveDomainsToDb 1`] = `
Array [
  Domain {
    "name": "www.cisa.gov",
    "webTechnologies": Array [
      Object {
        "categories": Array [
          Object {
            "id": 1,
            "name": "CMS",
            "priority": 1,
            "slug": "cms",
          },
        ],
        "confidence": 100,
        "cpe": "cpe:/a:drupal:drupal",
        "icon": "Drupal.svg",
        "name": "Drupal",
        "slug": "drupal",
        "version": "8",
        "website": "https://drupal.org",
      },
      Object {
        "categories": Array [
          Object {
            "id": 22,
            "name": "Web servers",
            "priority": 8,
            "slug": "web-servers",
          },
        ],
        "confidence": 100,
        "cpe": "cpe:/a:apache:http_server",
        "icon": "Apache.svg",
        "name": "Apache",
        "slug": "apache",
        "version": "",
        "website": "http://apache.org",
      },
      Object {
        "categories": Array [
          Object {
            "id": 27,
            "name": "Programming languages",
            "priority": 5,
            "slug": "programming-languages",
          },
        ],
        "confidence": 100,
        "cpe": "cpe:/a:php:php",
        "icon": "PHP.svg",
        "name": "PHP",
        "slug": "php",
        "version": "",
        "website": "http://php.net",
      },
    ],
  },
]
`;
